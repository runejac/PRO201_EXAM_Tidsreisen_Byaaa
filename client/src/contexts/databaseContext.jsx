import React, { createContext } from "react";
import { fetchJSON_client } from "../helpers/http.jsx";

export const DatabaseContext = createContext({
  async listHistory() {
    return await fetchJSON_client("/api/history");
  },
  async listQuiz(query) {
    return await fetchJSON_client("/api/quiz?" + new URLSearchParams(query));
  },
  async listAudio() {
    return await fetchJSON_client("/api/sound");
  },
});
