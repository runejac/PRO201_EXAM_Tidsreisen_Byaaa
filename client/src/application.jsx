import { BrowserRouter, Route, Routes } from "react-router-dom";
import { LoginPage } from "./pages/LoginPage.jsx";
import { MapPage } from "./pages/MapPage.jsx";
import { Quiz } from "./pages/capsules/quizCapsule/Quiz.jsx";
import { History } from "./pages/capsules/HistoryCapsule/History.jsx";
import { MyFindings } from "./pages/MyFindings.jsx";
import { UserContext } from "./contexts/userContext.jsx";
import { Profile } from "./pages/Profile.jsx";
import { NotFound } from "./components/NotFound.jsx";
import { BackButton } from "./components/BackButton";
import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useLoading } from "./helpers/useLoading.jsx";
import WithoutNavbar from "./components/WithoutNavbar";
import WithNavbar from "./components/WithNavbar";
import Audio from "./pages/capsules/AudioCapsule/Audio.jsx";
import Camera from "./components/Camera";
import WithBackButton from "./components/WithBackButton.jsx";

export const User = createContext("");
export const MapContext = React.createContext(null);

export function Application() {
  const [user, setUser] = useState([]);
  const [map, setMap] = useState(null);

  const { getUser } = useContext(UserContext);
  const { data: username, reload, loading, error } = useLoading(getUser);

  const providerValue = useMemo(() => ({ user, setUser }), [user, setUser]);
  const mapProvider = useMemo(() => ({ map, setMap }), [map, setMap]);

  useEffect(() => {
    username ? setUser(...username, username) : reload();
  }, [username]);

  return (
    <User.Provider value={providerValue}>
      <MapContext.Provider value={mapProvider}>
        <BrowserRouter>
          <main>
            <Routes>
              <Route element={<WithoutNavbar />}>
                <Route path="/" element={<LoginPage />} />
              </Route>
              <Route element={<WithNavbar />}>
                <Route path="/map" element={<MapPage />} />
                <Route path="/camera" element={<Camera />} />
                <Route element={<WithBackButton />}>
                  <Route path="/quiz/:id" element={<Quiz />} />
                  <Route path="/history/:id" element={<History />} />
                  <Route
                    path="/myfindings"
                    element={<MyFindings loading={loading} error={error} />}
                  />
                  <Route path="/audio/:id" element={<Audio />} />
                  <Route path="/profile" element={<Profile />} />
                  <Route path={"*"} element={<NotFound />} />
                </Route>
              </Route>
            </Routes>
          </main>
        </BrowserRouter>
      </MapContext.Provider>
    </User.Provider>
  );
}
