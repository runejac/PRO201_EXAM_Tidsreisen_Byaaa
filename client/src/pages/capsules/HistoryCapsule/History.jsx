import React, { useContext, useState } from "react";
import { DatabaseContext } from "../../../contexts/databaseContext.jsx";
import { useLoading } from "../../../helpers/useLoading.jsx";
import { Loading } from "../../../components/Loading.jsx";
import { StoryCard } from "./StoryCard.jsx";
import { NotLoggedIn } from "../../../components/NotLoggedIn.jsx";
import { User } from "../../../application.jsx";
import { ErrorModal } from "../../../components/ErrorModal.jsx";
import sawEffect from "../../../assets/images/saw-effect.png";
import "./storycard.css";
import "../../../css/olaf.css";

export function History() {
  const { user } = useContext(User);
  const { name } = user;

  const [error, setEror] = useState();
  const { listHistory } = useContext(DatabaseContext);
  const {
    data,
    error: errorHistory,
    loading,
  } = useLoading(async () => await listHistory());

  if (loading) {
    return <Loading />;
  }

  if (errorHistory) {
    setEror(errorHistory);
    console.error(error);
    return <ErrorModal error={error} />;
  }

  if (name === undefined) {
    return <NotLoggedIn />;
  }

  return (
    <div
      style={{
        backgroundColor: "#F2F1E8",
        height: "100vh",
      }}
    >
      <img
        id={"image-saw-effect"}
        height={700}
        src={sawEffect}
        alt="bilde av vann sag effekt bakgrunn"
      />
      {data && (
        <div>
          {data?.map((historyCapsule, index) => {
            return (
              <div key={index}>
                <StoryCard
                  loading={loading}
                  error={error}
                  historyCapsule={historyCapsule}
                />
              </div>
            );
          })}
        </div>
      )}
    </div>
  );
}
