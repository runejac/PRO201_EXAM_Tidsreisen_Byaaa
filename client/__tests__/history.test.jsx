/**
 * @jest-environment jsdom
 */

import React from "react";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";

import { User } from "../src/application.jsx";
import { DatabaseContext } from "../src/contexts/databaseContext.jsx";
import { History } from "../src/pages/capsules/HistoryCapsule/History.jsx";

//TODO: Se linje 29-54 i History.jsx

describe("History", () => {
  it("shows snapshot", async () => {
    const user = {
      name: "Hei",
      points: 23,
      finishedCapsules: [],
      intro: true,
      walk: false,
    };
    const context = jest.fn();
    let component;

    component = renderer.create(
      <MemoryRouter>
        <DatabaseContext.Provider value={{ context }}>
          <User.Provider value={{ user }}>
            <History />
          </User.Provider>
        </DatabaseContext.Provider>
      </MemoryRouter>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
