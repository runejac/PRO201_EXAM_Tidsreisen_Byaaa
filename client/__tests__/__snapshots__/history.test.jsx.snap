// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History shows snapshot 1`] = `
<div
  className="d-flex position-absolute w-100 vh-100 justify-content-center"
>
  <div
    className="spinner-border align-self-center"
    role="status"
  />
</div>
`;
